### PR TITLE
Add support for httplug

### LINKED ISSUE #107: Add support for httplug
Instead of relying on Guzzle 5 and lock deps down to this version (see https://github.com/j0k3r/graby/issues/8), it should be better to add support for [httplug](https://github.com/php-http/httplug) to be able to support multiple Guzzle version (or even other http lib).

### PR DESCRIPTION
**Pull request by aaa2000**
Close #107 

- [x] Fix failed tests and skipped test `testImageFile`
- [x] Update composer dependencies to a stable version
- [x] Check on a wallabag instance

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -26,12 +25,22 @@
         "monolog/monolog": "^1.13.1",
         "smalot/pdfparser": "~0.9.24",
         "wallabag/tcpdf": "^6.2",
-        "true/punycode": "~2.0"
+        "true/punycode": "~2.0",
+        "psr/http-message": "^1.0",
+        "php-http/client-implementation": "^1.0",
+        "php-http/httplug": "^1.0",
+        "php-http/message-factory": "^1.0",
+        "php-http/discovery": "^1.0",
```
Comment (jtojnar) [inline composer.json]: This will require PHP 5.5.

Comment (j0k3r) [inline composer.json]: This is fine to me.
Jumping to httplug will force a new major release (ie 2.0.0) so we can increase the minimum version of PHP in that case.

File: README.md [hunk 2]
```diff
@@ -35,7 +35,10 @@ That's why I made this fork:
 
 Add the lib using composer:
 
-    composer require j0k3r/graby
+    composer require j0k3r/graby php-http/guzzle5-adapter
+
+Why `php-http/guzzle6-adapter`? Because Graby is decoupled form any HTTP messaging client with help by [HTTPlug](http://httplug.io/).
```
Comment (j0k3r) [inline README.md]: Shouldn't this be `php-http/guzzle5-adapter`?
Or the line above is wrong and should be `php-http/guzzle6-adapter`

File: src/Extractor/HttpClient.php [hunk 3]
```diff
@@ -115,39 +139,34 @@ public function fetch($url, $skipTypeVerification = false, $httpHeader = [])
...
+            /** @var ResponseInterface $response */
             $response = $this->client->$method(
                 $url,
                 [
-                    'headers' => [
-                        'User-Agent' => $this->getUserAgent($url, $httpHeader),
-                        // add referer for picky sites
-                        'Referer' => $this->getReferer($url, $httpHeader),
-                    ],
-                    'timeout' => $this->config['timeout'],
```
Comment (j0k3r) [inline src/Extractor/HttpClient.php]: We can't configure timeout anymore?

File: tests/Extractor/HttpClientTest.php [hunk 4]
```diff
@@ -568,39 +327,21 @@ public function testTimeout()
...
-
-        $client = $this->getMockBuilder('GuzzleHttp\Client')
-            ->disableOriginalConstructor()
-            ->getMock();
+        $httpMockClient = new HttpMockClient();
 
-        $client->expects($this->any())
-            ->method('get')
-            ->willReturn($response);
+        for ($i = 0; $i < 10; $i++) {
```
Comment (j0k3r) [inline tests/Extractor/HttpClientTest.php]: Why will it be called 10 times?

File: tests/GrabyTest.php [hunk 5]
```diff
@@ -609,66 +418,52 @@ public function testAssetExtensionTXT()
...
-
-        $client = $this->getMockBuilder('GuzzleHttp\Client')
-            ->disableOriginalConstructor()
-            ->getMock();
-
-        $client->expects($this->any())
-            ->method('get')
-            ->willReturn($response);
+            ],
+            <<<"HTML"
```
Comment (j0k3r) [inline tests/GrabyTest.php]: :+1:

File: composer.json (hunk 1 above)
Comment (jtojnar) [inline composer.json]: Oh, did not notice Wallabag is switching to PHP 5.6, at selfoss we still support [5.4](https://wiki.debian.org/PHP#Available_versions) at least until Debian wheezy [EOL](https://wiki.debian.org/LTS).

Comment (aaa2000) [inline composer.json]: Add PHP 5.5 requirement

Comment (aaa2000) [inline composer.json]: I see that `j0k3r/safecurl` will no longer be used, I must analyse his behaviour to reimplement with  `httplug`

File: src/Extractor/HttpClient.php (hunk 3 above)
Comment (aaa2000) [inline src/Extractor/HttpClient.php]: Unfortunately, not easily. Each adapter has a different configuration. We must configure a client manually and inject it

File: tests/Extractor/HttpClientTest.php (hunk 4 above)
Comment (aaa2000) [inline tests/Extractor/HttpClientTest.php]: The configuration of redirection by httplug was incorrect and the test was not well implemented, I changed it. Is it better ?

File: README.md (hunk 2 above)
Comment (aaa2000) [inline README.md]: Fix

**Comment by aaa2000** [comment]
@j0k3r I wish to test temporarily with a fork of httplug but I have the error

```
Failed to clone the git@github.com:aaa2000/client-common.git repository, try running in interactive mode so that you can enter your GitHub credentials
                                                                                                                                                               
  [RuntimeException]                                                                                                                                            

  Failed to execute git clone --mirror 'git@github.com:aaa2000/client-common.git' '/home/travis/.composer/cache/vcs/git-github.com-aaa2000-client-common.git/'  
``` 

Do you know how to correct ?

**Comment by j0k3r** [comment]
Looks like it was a temporary hit limit I've relaunch build and it works fine now.

**Comment by j0k3r** [comment]
@aaa2000 if you want to join [Gitter](https://gitter.im/j0k3r/graby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=body_badge) we could talk about that

File: src/HttpClient/Plugin/ServerSideRequestForgeryProtection/Options.php [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+            }
+
+            return $this->{$list}[$type];
+        }
+
+        return $this->{$list};
+    }
+
+    /**
+     * Sets a list to the passed in array.
```
Comment (j0k3r) [inline src/HttpClient/Plugin/ServerSideRequestForgeryProtection/Options.php]: to the passed what?

File: tests/GrabyTest.php [hunk 7]
```diff
@@ -418,24 +419,33 @@ public function testAssetExtensionTXT()
...
     }
 
     /**
+     * @group dns-sensitive
      * @dataProvider dataForSinglePage
      */
     public function testSinglePage($url, $expectedUrl, $singlePageUrl)
     {
+        DnsMock::withMockedHosts([
+            'singlepage1.com' => [['type' => 'A', 'ip' => '93.184.216.34']],
```
Comment (j0k3r) [inline tests/GrabyTest.php]: Is it some random IP?

File: src/HttpClient/Plugin/ServerSideRequestForgeryProtection/Options.php (hunk 6 above)
Comment (aaa2000) [reply src/HttpClient/Plugin/ServerSideRequestForgeryProtection/Options.php]: Copy paste from safecurl, I will fix it later

File: tests/GrabyTest.php (hunk 7 above)
Comment (aaa2000) [reply tests/GrabyTest.php]: IP from example.com

Comment (j0k3r) [reply tests/GrabyTest.php]: Maybe you should put it in a constant to avoid XX updates if example.com changes its dns

Comment (aaa2000) [reply tests/GrabyTest.php]: Others are random ip. `93.184.216.34`  is `example.com`. The test use a http mock client, so a change doesn't impact the test. I will add a constant for the IP, it improves readability

File: src/HttpClient/Plugin/ServerSideRequestForgeryProtection/ServerSideRequestForgeryProtectionPlugin.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+class ServerSideRequestForgeryProtectionPlugin implements Plugin
+{
+    private $options;
+    private $uriFactory;
+
+    public function __construct(Options $options = null, UriFactory $uriFactory = null)
+    {
+        $this->options = $options ?: new Options();
+        $this->uriFactory = $uriFactory ?: UriFactoryDiscovery::find();
+        //TODO force using IPV4 ? @see SafeCurl::init()
```
Comment (aaa2000) [inline src/HttpClient/Plugin/ServerSideRequestForgeryProtection/ServerSideRequestForgeryProtectionPlugin.php]: I don't think that It is possible to do that with php-http

File: src/HttpClient/Plugin/ServerSideRequestForgeryProtection/ServerSideRequestForgeryProtectionPlugin.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+use Http\Message\UriFactory;
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Validates each part of the URL against a white or black list, to help protect against Server-Side Request Forgery
+ * attacks.
+ *
+ * @see https://github.com/j0k3r/safecurl
+ * @see https://whitton.io/articles/safecurl-ssrf-protection-and-a-capture-the-bitcoins/
+ */
```
Comment (aaa2000) [inline src/HttpClient/Plugin/ServerSideRequestForgeryProtection/ServerSideRequestForgeryProtectionPlugin.php]: Maybe, the plugin should be extract in his own repository

File: src/HttpClient/Plugin/ServerSideRequestForgeryProtection/ServerSideRequestForgeryProtectionPlugin.php (hunk 8 above)
Comment (j0k3r) [reply src/HttpClient/Plugin/ServerSideRequestForgeryProtection/ServerSideRequestForgeryProtectionPlugin.php]: Maybe we don't need it 🤔

File: src/HttpClient/Plugin/ServerSideRequestForgeryProtection/ServerSideRequestForgeryProtectionPlugin.php (hunk 9 above)
Comment (j0k3r) [reply src/HttpClient/Plugin/ServerSideRequestForgeryProtection/ServerSideRequestForgeryProtectionPlugin.php]: Might a good idea.

File: src/Extractor/HttpClient.php (hunk 3 above)
Comment (j0k3r) [reply src/Extractor/HttpClient.php]: I think this should be defined somewhere in the README

File: composer.json [hunk 10]
```diff
@@ -26,12 +30,23 @@
...
+        "php-http/client-common": "^1.6@dev",
+        "php-http/message": "^1.6@dev",
```
Comment (j0k3r) [inline composer.json]: Do you really need the `@dev`?

Comment (aaa2000) [reply composer.json]: Not anymore

File: src/Extractor/HttpClient.php (hunk 3 above)
Comment (aaa2000) [reply src/Extractor/HttpClient.php]: There is a paragraph `Timeout configuration` in the README

**Comment by j0k3r** [comment]
👏 !

**Pull request merged**